### ISSUE TITLE
Adding a random seed parameter for tensorflow.layer.dense

### ISSUE THREAD

**Issue by rasbt** [tensorflow/tensorflow#9744]
Hi,
I noticed that the `tensorflow.layer.dense` wrapper does not have a `seed` parameter, and I was wondering what you think about adding one. E.g., this seed would be used to initialize the random weights and bias units (if they are not initialized to zero).

Related to the points above,  it is currently also not clear how the weights are initialized. I.e., what distribution the random numbers are drawn from (if not zero). Does someone have some info on this? -- maybe this could be added to the API docs. 

What do you think? (I am happy to contribute a random seed implementation and a doc update if that's desirable.)

**Comment by concretevitamin** [comment]
Short answer to the seed request: it is unlikely this will be accepted, since TensorFlow's universal convention has been to use "initializers" instead of a simple seed.  For example, see [init_ops.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/init_ops.py).

That said, we should perhaps document the default inits used in `dense()` docstring.  @fchollet - mind taking a look?

**Comment by fchollet** [comment]
The way to get deterministic weight initializations in `dense` is to pass a `kernel_initializer` argument set to an initializer (instance of a class from `init_ops`) that has a fixed `seed`.

If `kernel_initializer` is not set, we use the global default weight initializer, which is `glorot_uniform`. But I think it would be good to have it be explicitly present in the `dense` signature instead of relying on an implicit global default, though.

**Comment by rasbt** [comment]
Thanks for the clarification. So the following would be the current default behavior based on the global default weight initializer?

```python
import tensorflow
from tensorflow.python.ops.init_ops import glorot_uniform_initializer

tensorflow.layers.dense(..., kernel_initializer=glorot_uniform_initializer())
```

I am happy to put together a PR to add a note to the docstrings of `Dense` and `dense`. When I understand correctly, you are suggesting to use `glorot_uniform_initializer` directly in `Dense`, i.e., via sth like

```python
if kernel_initializer is None:
    self.kernel_initializer = glorot_uniform_initializer()
else:
    self.kernel_initializer = kernel_initializer
```

Not sure what the current TensorFlow convention is; would it be okay to do these checks & assignments in the class `__init__`s directly?

**Comment by fchollet** [comment]
This code snippet in `__init__` would change neither the behavior of the layer nor its docstring and signature.  For the time being I would simply mention the default initializer in the docstring.

**Comment by rasbt** [comment]
Okay, will open a PR tomorrow.

> This code snippet in __init__ would change neither the behavior of the layer nor its docstring

I saw that in `Dense`'s `__init__` and the `dense` function, the bias initializer is set manually as well:

```python
def dense(
    inputs, units,
    activation=None,
    use_bias=True,
    kernel_initializer=None,
    bias_initializer=init_ops.zeros_initializer(),
    kernel_regularizer=None,
    bias_regularizer=None,
    activity_regularizer=None,
    trainable=True,
    name=None,
    reuse=None):
```

I am curious why setting `kernel_initializer=init_ops.glorot_uniform_initializer()` there wouldn't have an effect? I couldn't find anything in `Dense` and `dense` that would overwrite this setting ...

**Comment by fchollet** [comment]
> I am curious why setting kernel_initializer=init_ops.glorot_uniform_initializer() there wouldn't have an effect?

We should avoid instantiating anything in a class constructor signature, because the instance becomes global to all instances of the class. There are no immediate side effects in this case but it is bad practice and potentially dangerous for the user. For the same reason `bias_initializer=init_ops.zeros_initializer()`  should be avoided, but it is there for legacy reasons.

We should instantiate the initializers in the `__init__` of the class (so as to create new objects for every layer instance rather than a shared object everywhere). Keras achieves this by using default `initializer` argument values that are serial, i.e. that are simply the names of the initializers, which are then instantiated in the `__init__`.

**Comment by rasbt** [comment]
Ah thanks, makes sense now!

> This code snippet in __init__ would change neither the behavior of the layer nor its docstring and signature. For the time being I would simply mention the default initializer in the docstring.

Yeah, previously, I think I misinterpreted this. I thought you meant that it wouldn't have an effect setting something in `__init__` but when I understand correctly now, it already uses glorot_uniform, this setting this in `__init__` would have no effect. Only reason I thought this might be useful is that it might not be clear for someone looking at the code what the default weight initializer is, but yeah, I think leaving it as is and documenting it in the docstring should be fine.

**Comment by girving** [comment]
Closing since specifying the initializer is the right solution.
